internal/repository/repository.go

```diff
@@ -228,13 +228,13 @@ func (r *Repository) SaveAndEncrypt(ctx context.Context, t restic.BlobType, data
 	// get buf from the pool
 	ciphertext := getBuf()
 
-	ciphertext = ciphertext[:0]
+	*ciphertext = (*ciphertext)[:0]
 	nonce := crypto.NewRandomNonce()
-	ciphertext = append(ciphertext, nonce...)
+	*ciphertext = append(*ciphertext, nonce...)
 	defer freeBuf(ciphertext)
 
 	// encrypt blob
-	ciphertext = r.key.Seal(ciphertext, nonce, data, nil)
+	*ciphertext = r.key.Seal(*ciphertext, nonce, data, nil)
```

Comment:
Shouldn't it be enough to move the `defer freeBuf(ciphertext)` line below the call to `r.key.Seal`?
It looks to me that the problem here is that the call to freeBuf puts the original `ciphertext` buffer back into the pool and not the larger one after sealing.

The effect should be similar if the `New` function for `sync.Pool` allocated a chunk with `chunker.MaxSize`?

Reply:
Good catch. If we postpone defer, the stats are:

old=master, new=later defer
```
name              old time/op    new time/op    delta
SaveAndEncrypt-8    36.1ms ± 2%    35.7ms ± 2%     ~     (p=0.123 n=10+10)

name              old speed      new speed      delta
SaveAndEncrypt-8   116MB/s ± 2%   117MB/s ± 2%     ~     (p=0.123 n=10+10)

name              old alloc/op   new alloc/op   delta
SaveAndEncrypt-8    21.1MB ± 0%    17.0MB ± 0%  -19.31%  (p=0.000 n=10+10)

name              old allocs/op  new allocs/op  delta
SaveAndEncrypt-8      82.0 ± 0%      81.0 ± 0%   -1.22%  (p=0.000 n=10+10)
```

old=no pool, new=later defer
```
name              old time/op    new time/op    delta
SaveAndEncrypt-8    36.5ms ± 2%    35.7ms ± 2%   -2.02%  (p=0.004 n=10+10)

name              old speed      new speed      delta
SaveAndEncrypt-8   115MB/s ± 2%   117MB/s ± 2%   +2.07%  (p=0.004 n=10+10)

name              old alloc/op   new alloc/op   delta
SaveAndEncrypt-8    21.0MB ± 0%    17.0MB ± 0%  -18.97%  (p=0.000 n=10+10)

name              old allocs/op  new allocs/op  delta
SaveAndEncrypt-8      81.0 ± 0%      81.0 ± 0%     ~     (all equal)
```

old=later defer, new=this PR
```
name              old time/op    new time/op    delta
SaveAndEncrypt-8    35.7ms ± 2%    35.4ms ± 1%  -1.06%  (p=0.033 n=10+7)

name              old speed      new speed      delta
SaveAndEncrypt-8   117MB/s ± 2%   119MB/s ± 1%  +1.06%  (p=0.033 n=10+7)

name              old alloc/op   new alloc/op   delta
SaveAndEncrypt-8    17.0MB ± 0%    17.0MB ± 0%  -0.03%  (p=0.029 n=10+10)

name              old allocs/op  new allocs/op  delta
SaveAndEncrypt-8      81.0 ± 0%      80.0 ± 0%  -1.23%  (p=0.000 n=10+10)
```

Allocating chunker.MaxSize in addition to later defer gives:

old=master, new=later defer+bigger alloc
```
name              old time/op    new time/op    delta
SaveAndEncrypt-8    36.1ms ± 2%    36.3ms ± 4%     ~     (p=0.631 n=10+10)

name              old speed      new speed      delta
SaveAndEncrypt-8   116MB/s ± 2%   116MB/s ± 4%     ~     (p=0.631 n=10+10)

name              old alloc/op   new alloc/op   delta
SaveAndEncrypt-8    21.1MB ± 0%    17.0MB ± 1%  -19.49%  (p=0.000 n=10+10)

name              old allocs/op  new allocs/op  delta
SaveAndEncrypt-8      82.0 ± 0%      81.0 ± 0%   -1.22%  (p=0.000 n=10+9)
```

old=later defer+bigger alloc, new=this PR
```
name              old time/op    new time/op    delta
SaveAndEncrypt-8    36.3ms ± 4%    35.4ms ± 1%  -2.63%  (p=0.005 n=10+7)

name              old speed      new speed      delta
SaveAndEncrypt-8   116MB/s ± 4%   119MB/s ± 1%  +2.66%  (p=0.005 n=10+7)

name              old alloc/op   new alloc/op   delta
SaveAndEncrypt-8    17.0MB ± 1%    17.0MB ± 0%    ~     (p=0.143 n=10+10)

name              old allocs/op  new allocs/op  delta
SaveAndEncrypt-8      81.0 ± 0%      80.0 ± 0%  -1.23%  (p=0.000 n=9+10)
```

In a single table:
```
name \ time/op    master        nopool        later-defer   later-defer-maxsize  pointers
SaveAndEncrypt-8   36.1ms ± 2%   36.5ms ± 2%   35.7ms ± 2%          36.3ms ± 4%   35.4ms ± 1%

name \ speed      master        nopool        later-defer   later-defer-maxsize  pointers
SaveAndEncrypt-8  116MB/s ± 2%  115MB/s ± 2%  117MB/s ± 2%         116MB/s ± 4%  119MB/s ± 1%

name \ alloc/op   master        nopool        later-defer   later-defer-maxsize  pointers
SaveAndEncrypt-8   21.1MB ± 0%   21.0MB ± 0%   17.0MB ± 0%          17.0MB ± 1%   17.0MB ± 0%

name \ allocs/op  master        nopool        later-defer   later-defer-maxsize  pointers
SaveAndEncrypt-8     82.0 ± 0%     81.0 ± 0%     81.0 ± 0%            81.0 ± 0%     80.0 ± 0%
```